### PR TITLE
Resolve only relative references in open files on syntax server

### PR DESCRIPTION
**Pull request by sheetalkamat**
Resolve only relative module references in the open files

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
Looks like there's no packed build because of quote preference changes in a fourslash test.

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

**Comment by sheetalkamat** [comment]
@andrewbranch  can you please review again. I had to add another commit to ensure we aren't included triple slash reference tree (but including them only from open file) since we removed `noResolve`. 

@DanielRosenwasser @RyanCavanaugh  what is your take on this. do we want to get this in?

**Comment by DanielRosenwasser** [comment]
Yes, I think we want to get this in, but I have kind of been worrying about two scenarios:

1. How well does this work when a file has 100+ imports? It's an edge-case, but it's something we should consider.
2. Is there a way to remove pressure from the project system over time and go back to the single-file view of the world? Once the semantic server is loaded, we probably can unload some memory.

File: src/services/services.ts [hunk 1]
```diff
@@ -1309,21 +1309,13 @@ namespace ts {
...
-
-            if (host.resolveModuleNames) {
-                compilerHost.resolveModuleNames = (...args) => host.resolveModuleNames!(...args);
-            }
-            if (host.resolveTypeReferenceDirectives) {
-                compilerHost.resolveTypeReferenceDirectives = (...args) => host.resolveTypeReferenceDirectives!(...args);
-            }
-            if (host.useSourceOfProjectReferenceRedirect) {
-                compilerHost.useSourceOfProjectReferenceRedirect = () => host.useSourceOfProjectReferenceRedirect!();
-            }
```
Comment (andrewbranch) [inline src/services/services.ts:1326]: ❤️

File: src/server/editorServices.ts [hunk 2]
```diff
@@ -2967,7 +2967,15 @@ namespace ts.server {
...
             let defaultConfigProject: ConfiguredProject | undefined;
             let retainProjects: ConfiguredProject[] | ConfiguredProject | undefined;
-            if (!project && !this.syntaxOnly) { // Checking syntaxOnly is an optimization
+            if (this.syntaxOnly) {
+                // Invalidate resolutions in the file since this file is now open
+                info.containingProjects.forEach(project => {
+                    if (project.resolutionCache.removeRelativeNoResolveResolutionsOfFile(info.path)) {
+                        project.markAsDirty();
+                    }
+                });
```
Comment (DanielRosenwasser) [inline src/server/editorServices.ts:2976]:
```suggestion
                for (const project of info.containingProjects) {
                    if (project.resolutionCache.removeRelativeNoResolveResolutionsOfFile(info.path)) {
                        project.markAsDirty();
                    }
                }
```

File: src/compiler/resolutionCache.ts [hunk 3]
```diff
@@ -741,6 +763,32 @@ namespace ts {
...
+                }
+            });
+            return invalidated;
+        }
+
+        function removeRelativeNoResolveResolutionsOfFile(filePath: Path) {
+            let invalidated = removeRelativeNoResolveResolutionsOfFileFromCache(resolvedModuleNames, filePath, noResolveResolvedModule);
+            invalidated = removeRelativeNoResolveResolutionsOfFileFromCache(resolvedTypeReferenceDirectives, filePath, noResolveResolvedTypeReferenceDirective) || invalidated;
+            return invalidated;
+
```
Comment (DanielRosenwasser) [inline src/compiler/resolutionCache.ts:789]:
```suggestion
```

File: src/compiler/program.ts [hunk 4]
```diff
@@ -2647,7 +2657,15 @@ namespace ts {
             return projectReferenceRedirects.get(projectReferencePath) || undefined;
         }
 
+        function includeTripleslashReferencesFrom(file: SourceFile) {
+            return !host.includeTripleslashReferencesFrom || host.includeTripleslashReferencesFrom(file.originalFileName);
+        }
+
         function processReferencedFiles(file: SourceFile, isDefaultLib: boolean) {
+            if (!includeTripleslashReferencesFrom(file)) {
+                (skippedTrippleSlashReferences || (skippedTrippleSlashReferences = new Set())).add(file.path);
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts:2666]:
```suggestion
                (skippedTrippleSlashReferences ||= new Set()).add(file.path);
```

**Comment by sheetalkamat** [comment]
> 1. How well does this work when a file has 100+ imports? It's an edge-case, but it's something we should consider.

Should we cap this to say resolving 10 module names per file or something. But i am not sure if anyone uses those many relative imports in the file. Dont have data to cap on number of say that we should handle it.

> 2. Is there a way to remove pressure from the project system over time and go back to the single-file view of the world? Once the semantic server is loaded, we probably can unload some memory.

Do you think we should only use single file as open file (file thats last opened or asked to update project based on semantic operation requested). I did consider this but that just means we wont reuse the program as often if say user navigates to new file using go to def and hovers over certain item and goes back as an example. The current one keeps program updates to minimal and keeps on reusing things till it can.. leading to less program updates and more info. I am open to suggestion though.

**Comment by DanielRosenwasser** [comment]
> Should we cap this to say resolving 10 module names per file or something

Let's do it in a later pass, I'd like to try to get this in the next nightly so that we can get feedback as soon as we can.

> The current one keeps program updates to minimal and keeps on reusing things till it can

Yeah, that's a good point. If this can unblock those strenuous cross-project program loads, I think the responsiveness will make up for it.

**Comment by sheetalkamat** [comment]
Ok then merging.

**Pull request merged**